### PR TITLE
PyArr V2.0.0

### PR DESCRIPTION
**Pull request by Archmonger**
Targeting the following
- Support for all Sonarr & Radarr endpoints
- Bugfix all current endpoints
- Use requests.session `params` argument
- URL encode safety for all params
- Reorganize and refactor code to better match Sonarr/Radarr API documentation
- Add docstrings to all functions and classes
- Improve function naming consistency
- Use explicit `args` instead of generic `**kwargs` when possible
- Meet `black` and `pep8` code formatting standards
- Handle all remaining TODOs
- Deprecate Radarr V1 API (no longer maintained by Radarr team, and will be removed soon with Radarr v4)

Note: Since this PyArr refactoring effort is pretty large, pyarr users should expect to perform at least some minor refactoring (function names and arg names). However, all previously supported functionality has remained.

### CODE REVIEW

**Comment by Archmonger** [comment]
Current issues that have been resolved
fix #71 
fix #69

**Comment by Archmonger** [comment]
Radarr refactoring and API wrappers are complete. Will now begin Sonarr refactoring.

**Comment by Archmonger** [comment]
Sonarr reorganization complete. I determined several endpoints still haven't been developed within pyarr
- GET /parse
- GET/POST /release
- GET/POST/PUT/DELETE /tag and /tag/{id}

Will develop these endpoints prior to refactoring.

**Comment by Archmonger** [comment]
All endpoints for both Sonarr and Radarr are available within Pyarr. Refactoring/debugging all Sonarr endpoints is the only task remaining. Will complete by Saturday.

@marksie1988 Will you be available on Sunday to code review this PR? Also, I will need support revising the docs as some point down the line.

**Comment by marksie1988** [comment]
> All endpoints for both Sonarr and Radarr are available within Pyarr. Refactoring/debugging SonarrAPI is is the only task remaining. Will complete by Saturday.
> 
> @marksie1988 Will you be available on Sunday to code review this PR? Also, I will need support revising the docs as some point down the line.

Hey, thanks for all the work here, haven't had time to do much on this project for a while. 

I'm going to be busy today but will try review if I get chance, if not I'll do it Monday instead. 

I have thought about swapping to swagger for the docs hopefully will get some time to look at that soon.

**Comment by Archmonger** [comment]
Let's aim for Monday. Ended up socializing today and didn't get to doing the last leg of Pyarr development.

**Comment by Archmonger** [comment]
Got sidetracked again so couldn't cleanup and bugfix the SonarrAPI yet, but I will ping you on this PR when complete.

**Comment by Archmonger** [comment]
@marksie1988 This PR is ready for review & potential merge.

This fixes all open issues besides https://github.com/totaldebug/pyarr/issues/67 and closes all open pull requests.

Since this is pretty close to a Pyarr rewrite, I recommend reading the commit titles and then manually inspecting the python files as a whole. Inspecting each commit individually would be fairly painful.

**Comment by Archmonger** [comment]
@marksie1988 Do you know when you'd have availability to review?

Also, just want to mention again I think it's worth considering using [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) to directly convert our source code into documentation for v2. Would require significantly less effort than maintaining docstrings and docs separately.

**Review by marksie1988 — COMMENTED** [review]
> @marksie1988 Do you know when you'd have availability to review?
> 
> Also, just want to mention again I think it's worth considering using [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) to directly convert our source code into documentation for v2. Would require significantly less effort than maintaining docstrings and docs separately.

Hey, 

I have started to review, I got about halfway through sonarr but then got busy, I'll hopefully finish it all tonight.

File: pyarr/sonarr_api.py [hunk 1]
```diff
@@ -1,470 +1,677 @@
...
+            None
+        Returns:
+            JSON Response (dict)
+        """
+        path = "/api/queue"
         res = self.request_get(path)
         return res
 
+    # DELETE /queue
+    def del_queue(self, id_, blacklist=True):
```
Comment (marksie1988) [inline pyarr/sonarr_api.py:307]: Blacklist should be default to false, only blacklist if explicitly requested.

File: pyarr/sonarr_api.py [hunk 2]
```diff
@@ -1,470 +1,677 @@
...
+        """
+        params = {"id": id_, "blacklist": blacklist}
+        path = "/api/queue/"
+        res = self.request_del(path, params=params)
+        return res
+
+    ## PARSE
+
+    # GET /parse
+    def get_parsed_title(self, title):
```
Comment (marksie1988) [inline pyarr/sonarr_api.py:325]: Error is displayed if a title is provided:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/marksie1988/repos/pyarr/pyarr/sonarr_api.py", line 335, in get_parsed_title
    res = self.request_get(path, params=params)
  File "/home/marksie1988/repos/pyarr/pyarr/request_api.py", line 51, in request_get
    return res.json()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 898, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3/dist-packages/simplejson/__init__.py", line 525, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python3/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```

File: pyarr/sonarr_api.py [hunk 3]
```diff
@@ -1,470 +1,677 @@
...
+        Returns:
+            JSON Response (dict)
+        """
+        params = {"title": title}
+        path = "/api/parse"
+        res = self.request_get(path, params=params)
+        return res
+
+    # GET /parse
+    def get_parsed_path(self, path):
```
Comment (marksie1988) [inline pyarr/sonarr_api.py:339]: same error as for the parsed_title

File: pyarr/radarr_api.py [hunk 4]
```diff
@@ -0,0 +1,998 @@
+from .request_api import RequestAPI
+
+
+class RadarrAPI(RequestAPI):
+    """API wrapper for Radarr endpoints."""
+
+    def construct_movie_json(
```
Comment (marksie1988) [inline pyarr/radarr_api.py:7]: should this be private as it a user would notmally just need to use add_movie without needing to construct the json first?

File: pyarr/radarr_api.py [hunk 5]
```diff
@@ -0,0 +1,998 @@
...
+        """
+        path = f"/api/v3/movie/{id_}"
+        res = self.request_del(path)
+        return res
+
+    ## HISTORY
+
+    # GET /history
+    def get_history(
+        self, page, page_size=20, sort_direction="descending", sort_key="date"
```
[the remark this answers is not in the document]
Comment (marksie1988) [reply pyarr/radarr_api.py:263]: also this seems to error even when a page is supplied: 

```
>>> print(radarr.get_history(1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/pyarr/pyarr/radarr_api.py", line 282, in get_history
    res = self.request_get(path, params=params)
  File "/pyarr/pyarr/request_api.py", line 51, in request_get
    return res.json()
  File "/root/.local/lib/python3.8/site-packages/requests/models.py", line 900, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

File: pyarr/radarr_api.py [hunk 6]
```diff
@@ -0,0 +1,998 @@
...
+        res = self.request_get(path, params=params)
+        return res
+
+    ## BLACKLIST
+
+    # GET /blacklist
+    def get_blacklist(
+        self,
+        page,
+        page_size=20,
```
Comment (marksie1988) [inline pyarr/radarr_api.py:308]: Set a default to match sonarr `page=1, page_size=10`

File: pyarr/radarr_api.py [hunk 7]
```diff
@@ -0,0 +1,998 @@
...
+        res = self.request_del(path, data=data)
+        return res
+
+    ## QUEUE
+
+    # GET /queue
+    def get_queue(
+        self,
+        page,
+        page_size=20,
```
Comment (marksie1988) [inline pyarr/radarr_api.py:378]: set defaults `page=1, page_size=10`

File: pyarr/radarr_api.py [hunk 8]
```diff
@@ -0,0 +1,998 @@
...
+        if not id_:
+            path = "/api/v3/tag"
+        else:
+            path = f"/api/v3/tag/{id_}"
+
+        res = self.request_get(path)
+        return res
+
+    # POST /tag
+    def create_tag(self, id_, label):
```
Comment (marksie1988) [inline pyarr/radarr_api.py:680]: id should be removed, when creating a tag no id is required: 

```
[{'propertyName': 'Id', 'errorMessage': "'Id' must be equal to '0'.", 'attemptedValue': 1, 'severity': 'error', 'errorCode': 'EqualValidator', 'formattedMessageArguments': [], 'formattedMessagePlaceholderValues': {'comparisonValue': 0, 'propertyName': 'Id', 'propertyValue': 1}, 'resourceName': 'EqualValidator'}]
```

**Review by marksie1988 — CHANGES_REQUESTED** [review]
All reviewed now, 

Keep the docstrings as google format, I am working on sphinx which will auto document everything.

**Comment by Archmonger** [comment]
Thanks for the review. Will try to find some time address these comments tomorrow.

File: pyarr/sonarr_api.py (hunk 2 above)
Comment (Archmonger) [reply pyarr/sonarr_api.py:325]: Unknown why this one fails - I'll need to test this in the future.

File: pyarr/sonarr_api.py (hunk 3 above)
Comment (Archmonger) [reply pyarr/sonarr_api.py:339]: Unknown why this one fails - I'll need to test this in the future.

File: pyarr/radarr_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Archmonger) [reply pyarr/radarr_api.py]: For now I have left page_size = 20 to follow Radarr API defaults on these endpoints. 

Let's discuss whether we should follow the the Radarr docs, or match Sonarr docs.

File: pyarr/radarr_api.py (hunk 6 above)
Comment (Archmonger) [reply pyarr/radarr_api.py]: For now I have left page_size = 20 to follow Radarr API defaults on these endpoints. 

Let's discuss whether we should follow the the Radarr docs, or match Sonarr docs.

File: pyarr/radarr_api.py (hunk 7 above)
Comment (Archmonger) [reply pyarr/radarr_api.py]: For now I have left page_size = 20 to follow Radarr API defaults on these endpoints. 

Let's discuss whether we should follow the the Radarr docs, or match Sonarr docs.

File: pyarr/radarr_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Archmonger) [reply pyarr/radarr_api.py]: Also, unsure why this one fails. I'll need to debug it.

**Comment by Archmonger** [comment]
Three comments are remaining open. Have been commented with FIXMEs and I'll take a look at it this week to figure out what's wrong.

The remaining issues are open due to the `page_size` default differences between Radarr/Sonarr.
Let's discuss which path we want to take regarding `page_size`. Here's our options:

1) Follow Radarr and Sonarr `page_size` defaults separately
2) Follow Radarr `page_size` of 20 for all Sonarr functions
3) Follow Sonarr `page_size` of 10 for all Radarr functions

Since they're just default values it doesn't impact usability of Pyarr regardless of which route we take. I personally was leaning towards option 1 when developing this, but I'll let you make the final call.

**Comment by marksie1988** [comment]
> Three issues have been commented with FIXMEs and I'll take a look at it this week to figure out what's wrong. No issue immediately stood out to me when inspecting the code.
> 
> The remaining issues are open due to the `page_size` default differences between Radarr/Sonarr.
> Let's discuss which path we want to take regarding `page_size`. Here's our options:
> 
> 1. Follow Radarr and Sonarr `page_size` defaults separately, based on their respective API docs
> 2. Follow Radarr's default `page_size` of 20 for all Sonarr functions
> 3. Follow Sonarr's default `page_size` of 10 for all Radarr functions
> 
> Since they're just default values it doesn't impact usability of Pyarr regardless of which route we take. I personally was leaning towards option 1 when developing this, but I'll let you make the final call.

My thinking with making the results the same is that if someone is using both sonarr and radarr they would likely expect that the defaults for our wrapper would be the same. I get both sides of this really but not sure either one is better. Happy to leave with option 1 to save a re-work for not much gain.

**Comment by Archmonger** [comment]
Just confirmed the parse endpoint should work as expected, but will fail if the title does not contain both a season and episode number in it.

For example
- The following will parse correctly: `Series.Title.S01E00.720p.HDTV-Sonarr`
- The following will fail with no response: `Series.Title.720p.HDTV-Sonarr`

Can you confirm this?

**Comment by Archmonger** [comment]
Also the Sonarr team has just notified me that the `/api/v3` endpoints are live, but I don't see any docs for it yet.

So it looks like I'll be re-refactoring the `SonarrAPI` soon 😢

File: pyarr/sonarr_api.py (hunk 2 above)
Comment (marksie1988) [reply pyarr/sonarr_api.py:328]: yes using the fill title as `episodetitle.S01E01.HDTV-720p` does work

File: pyarr/radarr_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Archmonger) [reply pyarr/radarr_api.py]: @marksie1988 Just confirmed this endpoint works with default values. Perhaps you had a misconfiguration while testing?

**Comment by Archmonger** [comment]
@marksie1988 Confirmed that radarr `get_history` works on my end as is. Can you confirm this? That should be the last remaining issue on this PR.

**Comment by marksie1988** [comment]
> @marksie1988 Confirmed that radarr `get_history` works on my end as is. Can you confirm this? That should be the last remaining issue on this PR.

I will test this now.

**Review by marksie1988 — APPROVED** [review]

**Comment by Archmonger** [comment]
Let me remove the fixme comment and we'll be ready for merge on this PR

**Pull request merged**